tools/permission_prompt_test.py

```diff
@@ -5,220 +5,189 @@
 import pty
 import select
 import subprocess
+import sys
+import time
 
-from util import build_path, executable_suffix
+from util import build_path, executable_suffix, green_ok, red_failed
 
 PERMISSIONS_PROMPT_TEST_TS = "tools/permission_prompt_test.ts"
 
+PROMPT_PATTERN = b'⚠️'
+FIRST_CHECK_FAILED_PATTERN = b'First check failed'
+PERMISSION_DENIED_PATTERN = b'PermissionDenied: permission denied'
+
 
 # This function is copied from:
 # https://gist.github.com/hayd/4f46a68fc697ba8888a7b517a414583e
 # https://stackoverflow.com/q/52954248/1240268
-def tty_capture(cmd, bytes_input):
+def tty_capture(cmd, bytes_input, timeout=5):
     """Capture the output of cmd with bytes_input to stdin,
     with stdin, stdout and stderr as TTYs."""
     mo, so = pty.openpty()  # provide tty to enable line-buffering
     me, se = pty.openpty()
     mi, si = pty.openpty()
     fdmap = {mo: 'stdout', me: 'stderr', mi: 'stdin'}
 
+    timeout_exact = time.time() + timeout
     p = subprocess.Popen(
         cmd, bufsize=1, stdin=si, stdout=so, stderr=se, close_fds=True)
     os.write(mi, bytes_input)
 
-    timeout = .04  # seconds
+    select_timeout = .04  #seconds
     res = {'stdout': b'', 'stderr': b''}
     while True:
-        ready, _, _ = select.select([mo, me], [], [], timeout)
+        ready, _, _ = select.select([mo, me], [], [], select_timeout)
         if ready:
             for fd in ready:
                 data = os.read(fd, 512)
                 if not data:
                     break
                 res[fdmap[fd]] += data
-        elif p.poll() is not None:  # select timed-out
+        elif p.poll() is not None or time.time(
+        ) > timeout_exact:  # select timed-out
             break  # p exited
```

Comment:
Does p not need to be killed in the time out case? Or does closing the fds have the same effect?

Reply:
I was a little worried there for a moment. I had to test this one to be sure. It does manage to terminate the sub process before it leaves the scope of `tty_capture`. I figure this:
```python
for fd in [si, so, se, mi, mo, me]:
        os.close(fd)  # can't do it sooner: it leads to errno.EIO error
```
Causes the process to exit if it hasn't already.